### PR TITLE
feat: automapping plugin

### PR DESCRIPTION
**Pull request by hansjhoffman**
# Automap Plugin

Initially designed for a headless workflow in mind and thus decisions were made in the v1 design to accommodate that use case over others. For example, the automap process will fail when not all fields possess the minimum required confidence level (either by the 'system' or a 'user'). The idea behind this is that customers' data is coming from a trusted source, in a predictable format with predictable means of identifying each Sheet/Field. Therefore, the "system" must be trained first which allows the plugin to safely bail requiring a HITL to complete. 

Includes:

- 2 confidence levels for mapping
- Plugin specific logs to help easily identify when mixed in with numerous plugins
- Debug mode for development since the plugin can fail in numerous ways making it hard to debug
- User defined failure callback

---

Screenshot 2023-06-29 at 12 50 41 PM

### Acceptance Criteria for v1

- [x] I can upload a single sheet file and it will automap (.CSV)
- [x] Handle failure when not all fields can be confidently mapped (via 'confident' or 'strict')
- [x] Allow match using ids or names.
- [x] If the auto-map fails in any case, allow the user to pass in a property on what happens on failure.
    1. If you don’t have a matching file name or target, it skips

### Usage

```ts
import { Client, FlatfileEvent } from "@flatfile/listener";
import { automap } from "@flatfile/plugin-automap";

export default function flatfileEventListener(listener: Client) {
  // listener.on("**", (event) => {
  //   console.log(`Received event: ${JSON.stringify(event, null, 2)}`);
  // });

  listener.use(
    automap({
      accuracy: "confident",
      debug: process.env.NODE_ENV === "dev",
      defaultTargetSheet: "Contact",
      matchFilename: /test.csv$/g,
      onFailure: (event: FlatfileEvent) => {
        console.error(
          `Please visit https://spaces.flatfile.com/space/${event.context.spaceId}/files?mode=import to manually import file.`,
        );
      },
    }),
  );
}
```

### Known issues

- File name changes in the UI will mistakenly update when mapping is abandoned due to a lack of confidence. This is just a small logic bug that doesn't "break" the plugin
- Internal mapping certainty is subject to change at any moment and this makes the plugin even less predictable than it already is

### CODE REVIEW

File: plugins/automap/README.md [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# @flatfile/plugin-automap
+
+The purpose of this plugin is to provide automapping for headless workflows. 
+
+`npm i @flatfile/plugin-automap`
+
+## Get Started
+
+Follow [this guide](https://flatfile.com/docs/plugins/transform/automap) to learn how to use the plugin.
```
Comment (hansjhoffman) [inline plugins/automap/README.md:9]: ❔ Will this be the right link?

File: plugins/automap/src/automap.plugin.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+ *
+ * @property {string} accuracy - match columns either by 'confident' (> 90% match) or 'exact' (100% match).
+ * @property {boolean} debug - show helpul messages useful for debugging (use intended for development).
+ * @property {string} defaultTargetSheet - exact sheet name to import data to.
+ * @property {RegExp} matchFilename - a regular expression to match specific files to perform automapping on.
+ * @property {Function} onFailure - callback to be executed when plugin bails.
+ * @property {string} targetWorkbook - specify destination Workbook id or name.
+ */
+export interface AutomapOptions {
+  readonly accuracy: "confident" | "exact";
```
Comment (hansjhoffman) [inline plugins/automap/src/automap.plugin.ts:28]: Chose 'confident' over 'remembered' b/c technically 'exact' is 'remembered' too. But I'm open to whatever

File: plugins/automap/package.json [hunk 3]
```diff
@@ -0,0 +1,38 @@
+{
+  "name": "@flatfile/plugin-automap",
+  "version": "0.0.1",
```
Comment (hansjhoffman) [inline plugins/automap/package.json:3]: ❔ Should we keep the version like this for a pre 1.0 release?

File: plugins/automap/package.json [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+  "description": "A plugin for auto mapping imported files when confidence levels are high.",
+  "registryMetadata": {
+    "status": "Coming soon",
+    "category": "schema"
+  },
+  "engines": {
+    "node": ">= 12"
```
Comment (hansjhoffman) [inline plugins/automap/package.json:10]: ❔ Other plugins use this. Is that accurate? I was thinking more like ">= 16" from personal experience

File: plugins/automap/README.md (hunk 1 above)
Comment (ashleygmulligan2) [reply plugins/automap/README.md:9]: actually, does transform feel wrong here? We could make a new section for automations: https://flatfile.com/docs/plugins/automation/automap

File: plugins/automap/package.json [hunk 5]
```diff
@@ -0,0 +1,38 @@
+{
+  "name": "@flatfile/plugin-automap",
+  "version": "0.0.1",
+  "description": "A plugin for auto mapping imported files when confidence levels are high.",
+  "registryMetadata": {
+    "status": "Coming soon",
+    "category": "schema"
+  },
```
[the remark this answers is not in the document]
Comment (ashleygmulligan2) [reply plugins/automap/package.json:8]: automation for category

[the remark this answers is not in the document]
Comment (ashleygmulligan2) [reply plugins/automap/package.json:8]: status isn't used anymore

File: plugins/automap/package.json (hunk 3 above)
Comment (ashleygmulligan2) [reply plugins/automap/package.json:3]: yes

File: plugins/automap/package.json (hunk 4 above)
Comment (ashleygmulligan2) [reply plugins/automap/package.json:10]: 16 yea

**Review by ashleygmulligan2 — APPROVED** [review]

**Pull request merged**